docs/miscellaneous.rst

```diff
@@ -221,6 +221,125 @@ Either add ``--libraries "Math:0x12345678901234567890 Heap:0xabcdef0123456"`` to
 
+*****************
+Contract Metadata
+*****************
+
+The Solidity compiler automatically generates an internal json file, the
+contract metadata, that contains information about the current contract.
+It can be used to query the compiler version, the sourcecode, the ABI
+and NatSpec documentation in order to more safely interact with the contract
+and to verify its source code.
+
+The compiler inserts a swarm hash of that file into the bytecode of each
+contract, so that you can retrieve the file in an authenticated way
+without having to resort to a centralized data provider.
+
+Specifically, the runtime code for a contract always starts with
+``push32 <metadata hash> pop``, so you can take a look at the 32 bytes starting at
+the second byte of the code of a contract.
+
+Of course, you have to publish the metadata file to swarm (or some other service)
+so that others can access it. The file can be output by using ``solc --metadata``.
+It will contain swarm references to the source code, so you have to upload
+all source files and the metadata.
+
+The metadata file has the following format. The example below is presented in a
+human-readable way. Properly formatted metadata should use quotes correctly,
+reduce whitespace to a minimum and sort the keys of all objects to arrive at a
+unique formatting.
+Comments are of course also not permitted and used here only for explanatory purposes.
+
+    {
+      // Required: The version of the metadata format
+      version: "1",
+      // Required: Source code language, basically selects a "sub-version"
+      // of the specification
+      language: "Solidity",
+      // Required: Details about the compiler, contents are specific
+      // to the language.
+      compiler: {
+        // Required for Solidity: Version of the compiler
+        version: "0.4.6+commit.2dabbdf0.Emscripten.clang",
+        // Optional: Hash of the compiler binary which produced this output
+        keccak256: "0x123..."
+      },
+      // Required: Compilation source files/source units, keys are file names
+      sources:
+      {
+        "myFile.sol": {
+          // Required: keccak256 hash of the source file
+          "keccak256": "0x123...",
+          // Required (unless "content" is used, see below): URL to the
+          // source file, protocol is more or less arbitrary, but a swarm
+          // URL is recommended
+          "url": "bzzr://56ab..."
+        },
+        "mortal": {
+          "keccak256": "0x123...",
+          // Required (unless "url" is used): literal contents of the source file
+          "content": "contract mortal is owned { function kill() { if (msg.sender == owner) selfdestruct(owner); } }"
+        }
+      },
+      // Required: Compiler settings
+      settings:
+      {
+        // Required for Solidity: Sorted list of remappings
```

Comment:
Command line arguments in `argv` would be useful for any compiler (this can be deferred to version 2 of this format).